File: ClientBankSWPython/controller/client_controller.py
import sys
import requests
from controller.general_contoller import get_db

db = get_db()

def list_all_clients():
    clients_ref = db.collection(u'clients')
    docs = clients_ref.stream()
    clients = []
    for doc in docs:
        clients.append(doc.to_dict())
    return clients

def insert_client(str_id_number, doc_type, name, birthday, email, phone_number, gender):
    id_number = int(str_id_number)
    
    new_client = {
        u"identificationNumber" : id_number,
        u"documetType" : doc_type,
        u"name" : name,
        u"birthday" : birthday,
        u"email" : email,
        u"phoneNumber" : phone_number,
        u"gender" : gender
    }
    
    doc_ref = db.collection(u'clients').document(str_id_number)
    doc_ref.set(new_client)
    return True

def find_client(str_id_number):
    doc_ref = db.collection(u'clients').document(str_id_number)
    doc = doc_ref.get()
    if doc.exists:
        return doc.to_dict()
    else:
        raise Exception("Cliente no encontrado")

def update_client(str_id_number, doc_type, name, birthday, email, phone_number, gender):   
    try:
        client_to_update = find_client(str_id_number)
        id_number = int(str_id_number)
        
        client_to_update["identificationNumber"] = id_number
        client_to_update["documetType"] = doc_type
        client_to_update["name"] = name
        client_to_update["birthday"] = birthday
        client_to_update["email"] = email
        client_to_update["phoneNumber"] = phone_number
        client_to_update["gender"] = gender
        
        doc_ref = db.collection(u'clients').document(str_id_number)
        doc_ref.set(client_to_update)
        return True
    except Exception as ex:
        raise ex

def delete_client(str_id_number):   
    try:
        client_to_delete = find_client(str_id_number)
        db.collection(u'clients').document(str_id_number).delete()
        return True
    except Exception as ex:
        raise ex

File: ClientBankSWNet/ClientBankSWNet/controller/ClientController.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Google.Cloud.Firestore;

namespace ClientBankSWNet.controller
{
    class ClientController
    {
        private static ClientController controller = null;
        private static FirestoreDb db;

        public ClientController()
        {
            db = GeneralController.DataBase;
        }

        public static ClientController Instance
        {
            get
            {
                if (controller == null)
                    controller = new ClientController();

                return controller;
            }
        }

        public async Task<List<Dictionary<string, object>>> ListAllClients()
        {
            List<Dictionary<string, object>> list = new List<Dictionary<string, object>>();

            try
            {
                CollectionReference collRef = db.Collection("clients");
                QuerySnapshot snapshot = await collRef.GetSnapshotAsync();

                foreach (DocumentSnapshot document in snapshot.Documents)
                {
                    Dictionary<string, object> documentDictionary = document.ToDictionary();
                    list.Add(documentDictionary);
                }
            }
            catch (Exception ex)
            {
                throw ex;
            }

            return list;
        }

        public async Task<bool> AddClient(String strIdNumber, String docType, String name, DateTime birthday, String email, String phoneNumber, String gender)
        {
            if (strIdNumber.Equals("") || docType.Equals("") || name.Equals("") || birthday == null ||
                email.Equals("") || phoneNumber.Equals("") || gender.Equals(""))
            {
                throw new Exception("Por favor diligencie correctamente todos los campos");
            }
            else
            {
                int identificationNumber;
                try
                {
                    identificationNumber = Int32.Parse(strIdNumber);
                }
                catch (Exception)
                {
                    throw new Exception("Número de identificación invalido.\nPor favor introducir un valor numérico");
                }

                Dictionary<string, object> newClient = new Dictionary<string, object>() {
                    {"identificationNumber", identificationNumber },
                    {"documetType" , docType },
                    {"name" , name },
                    {"birthday" , birthday.ToString("yyyy-MM-dd") },
                    {"email" , email},
                    {"phoneNumber" , phoneNumber},
                    {"gender" , gender}
                };

                DocumentReference docRef = db.Collection("clients").Document(strIdNumber);
                await docRef.SetAsync(newClient);

                return true;
            }
        }

        public async Task<Dictionary<string, object>> FindClient(string strIdNumber)
        {
            if (strIdNumber.Equals(""))
            {
                throw new Exception("Por favor ingrese un número de identificación.");
            }
            else
            {
                int identificationNumber;
                try
                {
                    identificationNumber = Int32.Parse(strIdNumber);
                }
                catch (Exception)
                {
                    throw new Exception("Número de identificación invalido.\nPor favor introducir un valor numérico");
                }

                DocumentReference docRef = db.Collection("clients").Document(strIdNumber);
                DocumentSnapshot snapshot = await docRef.GetSnapshotAsync();

                if (snapshot.Exists)
                {
                    Dictionary<string, object> client = snapshot.ToDictionary();
                    return client;
                }
                else
                {
                    throw new Exception("El cliente con el número de identificación " + identificationNumber +
                    " no ha sido encontrado");
                }

            }
        }

        public async Task<bool> UpdateClient(String strIdNumber, String docType, String name, DateTime birthday, String email, String phoneNumber, String gender)
        {
            Dictionary<string, object> clientToUpdate;
            try
            {
                clientToUpdate = await this.FindClient(strIdNumber);
            }
            catch (Exception ex)
            {
                throw ex;
            }

            if (docType.Equals("") || name.Equals("") || birthday == null ||
                email.Equals("") || phoneNumber.Equals("") || gender.Equals(""))
            {
                throw new Exception("Por favor diligencie correctamente todos los campos");
            }
            else
            {
                int identificationNumber = Int32.Parse(strIdNumber);
                clientToUpdate = new Dictionary<string, object>() {
                    {"identificationNumber", identificationNumber },
                    {"documetType" , docType },
                    {"name" , name },
                    {"birthday" , birthday.ToString("yyyy-MM-dd") },
                    {"email" , email},
                    {"phoneNumber" , phoneNumber},
                    {"gender" , gender}
                };

                //Tener en cuenta
                DocumentReference docRef = db.Collection("clients").Document(strIdNumber);
                await docRef.SetAsync(clientToUpdate, SetOptions.MergeAll);

                return true;
            }
        }

        public async Task<bool> DeleteClient(String strIdNumber)
        {
            Dictionary<string, object> clientToDelete;
            try
            {
                clientToDelete = await this.FindClient(strIdNumber);
            }
            catch (Exception ex)
            {
                throw ex;
            }

            DocumentReference docRef = db.Collection("clients").Document(strIdNumber);
            await docRef.DeleteAsync();

            return true;
        }
    }
}
